river/drift/adwin_c.pyx

```diff
@@ -0,0 +1,379 @@
+
+    def get_total(self):
+        return self.total
+
+    def get_variance(self):
+        return self.variance
+
+    @property
+    def variance_in_window(self):
+        return self.variance / self.width
+
+    cdef void _init_buckets(self):
+        """ Initialize the bucket's list and statistics
+
+        Set all statistics to 0 and create a new bucket List.
+
+        """
+        self.bucket_deque: Deque['Bucket'] = deque([Bucket()])
+        self.total = 0.0
+        self.variance = 0.0
+        self.width = 0.0
+        self.n_buckets = 0
+
+    def update(self, value: float):
+        """Update the change detector with a single data point.
+
+        Apart from adding the element value to the window, by inserting it in
+        the correct bucket, it will also update the relevant statistics, in
+        this case the total sum of all values, the window width and the total
+        variance.
+
+        Parameters
+        ----------
+        value
+            Input value
+
+        Returns
+        -------
+        bool
+            If True then a change is detected.
+
+        """
+        return self._update(value)
+
+    cdef bint _update(self, double value):
+
+        cdef double incremental_variance
+
+        self.width += 1
+
+        self._insert_element(value, 0.0)
+        incremental_variance = 0.0
+        if self.width > 1.0:
+            incremental_variance = (
+                    (self.width - 1.0)
+                    * (value - self.total / (self.width - 1.0))
+                    * (value - self.total / (self.width - 1.0))
+                    / self.width
+            )
+        self.variance += incremental_variance
+        self.total += value
+        self._compress_buckets()
+
+        return self._detect_change()
+
+    cdef void _insert_element(self, double value, double variance):
+        cdef Bucket bucket = self.bucket_deque[0]
+        bucket.insert_data(value, variance)
+        self.n_buckets += 1
+
+        if self.n_buckets > self.max_n_buckets:
+            self.max_n_buckets = self.n_buckets
+
+    cdef double _bucket_size(self, int row):
+        return pow(2, row)
+
+    cdef double _delete_element(self):
+        """Delete an item from the bucket deque.
+
+        Deletes the last item and updates relevant statistics kept by ADWIN.
+
+        Returns
+        -------
+        The size of the item being dropped
+
+        """
+        cdef:
+            double n1, u1
+            Bucket bucket
+            double incremental_variance
+
+        bucket = self.bucket_deque[-1]
+        n1 = self._bucket_size(len(self.bucket_deque) - 1)
+        self.width -= n1
+        self.total -= bucket.get_total_at(0)
+        u1 = bucket.get_total_at(0) / n1
+        incremental_variance = (
+                bucket.get_variance_at(0) + n1 * self.width
+                * (u1 - self.total / self.width) * (u1 - self.total / self.width)
+                / (n1 + self.width)
+        )
```

Comment:
I am not aware of the this update rule for incremental variance estimators. It resembles (somewhat) what use in the stats module, but there are some portions that are alien to me.

Reply:
I have added a note in the docstring for the corresponding paper.